File: java/_554_Brick_Wall.java
import java.util.HashMap;
import java.util.List;
import java.util.Set;



public class _554_Brick_Wall {
    public int leastBricks(List<List<Integer>> wall) {
    	int tag=0;
    	HashMap<Integer, Integer>map=new HashMap<>();
    	for(int i=0;i<wall.size();i++) {
    		int sum=0;
    		List<Integer>row=wall.get(i);
    		int blocks=row.size();
    		if(blocks!=1)
    			tag=1;
    		for(int j=0;j<blocks-1;j++) {
    			sum+=row.get(j);
    			map.put(sum,map.getOrDefault(sum, 0)+1);
    		}
    	}
    	if(tag==0)
    		return wall.size();
    	Set<Integer>keys=map.keySet();
    	int max=Integer.MIN_VALUE;
    	for(Integer key:keys) {
    		int tmp=map.get(key);
    		if(tmp>max) 
    			max=tmp;
    	}
    	return wall.size()-max;
    }
}

File: cpp/554.Brick Wall.cpp
#include <hash_map>
#include <vector>
#include <unordered_map>
using namespace __gnu_cxx;
using namespace std;

class Solution {
public:
    int leastBricks(vector<vector<int>>& wall) {
        int height = wall.size();
        int tag = 0;
        unordered_map<int, int> map;
        for (int j=0; j<wall[0].size(); j++)
        	map[j] = 0;
        for (int i=0; i<height; i++){
        	int width = wall[i].size();
        	if (width != 1)
        		tag = 1;
        	int position = 0;
        	for (int j = 0; j < width-1; j++){
        		position += wall[i][j];
        		map[position] = map[position] + 1;
			}
		}
        if (tag == 0)
        	return height;
        int max = INT_MIN;
        for (auto& iter: map){
        	if (iter.second > max)
        		max = iter.second;
		}
		return height-max;
    }
};
